sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/MultimapUserState.java

```diff
@@ -0,0 +1,326 @@
+        !isClosed,
+        "Multimap user state is no longer usable because it is closed for %s",
+        keysStateRequest.getStateKey());
+
+    Collection<V> pendingValues = pendingAdds.get(key);
+    if (isCleared || negativeCache.contains(key) || pendingRemoves.contains(key)) {
+      return Iterables.limit(Iterables.unmodifiableIterable(pendingValues), pendingValues.size());
+    }
+
+    Iterable<V> persistedValues = getPersistedValues(key);
+    if (Iterables.isEmpty(persistedValues)) {
+      negativeCache.add(key);
+    }
+    return Iterables.concat(
+        persistedValues,
+        Iterables.limit(Iterables.unmodifiableIterable(pendingValues), pendingValues.size()));
+  }
+
+  @SuppressWarnings({
+    "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-12687)
+  })
+  /*
+   * Returns an iterables containing the key from each key-value pair in this multimap, without collapsing duplicates.
+   */
+  public Iterable<K> keys() {
+    checkState(
+        !isClosed,
+        "Multimap user state is no longer usable because it is closed for %s",
+        keysStateRequest.getStateKey());
+
+    if (isCleared) {
+      return Iterables.limit(Iterables.unmodifiableIterable(pendingKeys), pendingKeys.size());
+    }
+
+    Iterable<K> persistedKeys = getPersistedKeys();
+    persistedKeys = Iterables.filter(persistedKeys, key -> !pendingRemoves.contains(key));
+    return Iterables.concat(
+        persistedKeys,
+        Iterables.limit(Iterables.unmodifiableIterable(pendingKeys), pendingKeys.size()));
+  }
+
+  /*
+   * Store a key-value pair in the multimap.
+   * Allows duplicate key-value pairs.
+   */
+  public void put(K key, V value) {
+    checkState(
+        !isClosed,
+        "Multimap user state is no longer usable because it is closed for %s",
+        keysStateRequest.getStateKey());
+    pendingAdds.put(key, value);
+    pendingKeys.add(key);
+  }
+
+  /*
+   * Removes all values for this key in the multimap.
+   */
+  public void remove(@NonNull K key) {
+    checkState(
+        !isClosed,
+        "Multimap user state is no longer usable because it is closed for %s",
+        keysStateRequest.getStateKey());
+    pendingAdds.removeAll(key);
+    pendingKeys.removeAll(Collections.singletonList(key));
+    pendingRemoves.add(key);
+  }
+
+  @SuppressWarnings({
+    "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-12687)
+  })
+  // Update data in persistent store
+  public void asyncClose() throws Exception {
+    checkState(
+        !isClosed,
+        "Multimap user state is no longer usable because it is closed for %s",
+        keysStateRequest.getStateKey());
+
+    // Nothing to persist
+    if (!isCleared && pendingRemoves.isEmpty() && pendingAdds.isEmpty()) {
+      isClosed = true;
+      return;
+    }
+
+    // Clear currently persisted key-values
+    if (isCleared) {
+      for (K key : getPersistedKeys()) {
+        beamFnStateClient.handle(
+            createUserStateRequest(key)
```

Comment:
Can't this be one clear request containing MultimapKeysUserState instead of enumerating all the keys?